### PR TITLE
OnChain order cancellation

### PR DESCRIPTION
**Pull request by shoom3301**
# Summary

## Changes
1. Button for selecting a cancellation type (on/off-chain). Only for wallets that support off-chain cancellation
2. Displaying of tx cost for on-chain cancellation (Tx in Goerli always UKNOWN, because we don't have a gas price here)
3. Added a link to a cancellation hash

image

| | |
|--------|--------|
| image | image |

> For code-reviewers: It will be easier to review the code by commits

  # To Test
  
### Smart contract wallet

1. Open CowSwap with a smart contract wallet (Gnosis Safe or other)
2. Create a limit order with a price above market
3. Click "Cancel order" in the context menu in the orders table
- [ ] there should NOT be an option to change the cancellation type
4. Click "Request cancellation"
- [ ] an order must be canceled using on-chain transaction

### Other wallets

1. Open CowSwap with a NOT smart contract wallet (Metamask or other)
2. Create a limit order with a price above market
3. Click "Cancel order" in the context menu in the orders table
- [ ] There should be an option to change the cancellation type (on-chain/off-chain)
4. Click "Request cancellation"
- [ ] an order must be canceled in accordance with selected type

### Eth-flow orders
- [ ] Canecalltion of eth-flow orders should work as before

### CODE REVIEW

**Review by anxolin — COMMENTED** [review]
I manage to cancel it, but still the UX needs to be a bit polished. 


Once canceled, it shouldn't show you the cancel button. 
image


~It didn't track the "canceling..." state. I would think that we should be "canceling..." when we already posted the TX. Also, maybe nice to have a way to see the cancelation tx?~
EDIT: on my second attempt, it did. Maybe I just didn't test correctly.

~One trivial implementation of the above is to add the tx to the list of tx we show. Although not sure that's the best UX. We could discuss this~

Not needed if we just handle the "Canceling..." state right. Although the part of showing the cancelation TX somewhere might be nice addition if its not hard.

~Also, when the state changed from "Open" to canceled, the UI showed "Canceling..." instead of "Canceled" (now should be final, not like in offchain cancelation that its possible to execute later)~

On my second attempt, it didn't happen. Maybe I made it up??

**Comment by shoom3301** [comment]
@elena-zh 
1. Off-chain cancellation is not a 100% guarantee. Potentially, an order might be filled even after off-chain cancellation. On-chain cancellation is 100% guaranteed.
2. Oops, will fix it
3. Yes, it's possible. Let's discuss it with the team. Anyway, we can improve it after merging this PR (I guess it should be fixed on the backend side)
4. Hmmm, I'm nott aware about this case, let's discuss it
5. Will check it
6. o_0 it shouldn't be possible
7. Out of scope

**Comment by shoom3301** [comment]
@elena-zh 
1. We discussed
2. Fixed
3. I couldn't reproduce. Anyway, I don't think we can fix it on frontend side
4. Fixed. Now you can't cancel an order while it in Signing state
5. I don't know how to implement it. Let's create another issue for it
6. Fixed
7. Out of scope

**Comment by elena-zh** [comment]
Hey @shoom3301 , 

1. How to reproduce case 3: see the [video](https://www.loom.com/share/18c2581b29324b06a3f8f6238e447b8f)
2. Still I can see 'Cancel' order button for Cancelled (offchain), Expired, Cancelling orders when connected to EOA
cancel for cancelled
3. I have noticed that when I sign offchain cancel order tx, the order goes to cancelling state, then can be returned into the Open state, then again can go to 'Cancelling' state. You can see it on the video in p1. 
4. I think, there is no need to show Cancel TX link for the ETH orders, as it presents on the progress bar. WDYT?
cancelling
4. Cancel order link (when connected to SC wallets) still navigates to the 404 page
not correct
https://goerli.etherscan.io/tx/0x3678d1b1b3ea016158f3aa053cf9d1a3f7ed26644823fb30d2ef229d2f58c692
Just a bit more context: maybe we could copy the link logic that is implemented for this component?
image


6. Cancellation TX link is still displayed for Filled orders (SC wallet)
6 not removed

Thanks

File: src/cow-react/common/hooks/useCancelOrder/useSendOnChainCancellation.test.ts [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+import { useWalletInfo } from '@cow/modules/wallet'
+import { useRequestOrderCancellation, useSetOrderCancellationHash } from 'state/orders/hooks'
+import { useTransactionAdder } from 'state/enhancedTransactions/hooks'
+import { useEthFlowContract, useGP2SettlementContract } from 'hooks/useContract'
+import { NATIVE_CURRENCY_BUY_TOKEN } from 'constants/index'
+import { BigNumber } from '@ethersproject/bignumber'
+import { COW } from 'constants/tokens'
+
+const chainId = 1
+const settlementCancellationTxHash = '0xcfwj23g4fwe111'
```
Comment (anxolin) [inline src/cow-react/common/hooks/useCancelOrder/useSendOnChainCancellation.test.ts:13]: Lol, it looks like you hit your keyboard hard to randomize this address

File: src/cow-react/common/hooks/useCancelOrder/useSendOnChainCancellation.test.ts [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+      expect(transactionAdder).toHaveBeenCalledTimes(1)
+      expect(transactionAdder.mock.calls[0]).toMatchSnapshot()
+      expect(requestOrderCancellation).toHaveBeenCalledTimes(1)
+      expect(requestOrderCancellation.mock.calls[0]).toMatchSnapshot()
+      expect(setOrderCancellationHash).toHaveBeenCalledTimes(1)
+      expect(setOrderCancellationHash.mock.calls[0]).toMatchSnapshot()
+    })
+  })
+})
```
Comment (anxolin) [inline src/cow-react/common/hooks/useCancelOrder/useSendOnChainCancellation.test.ts:120]: Vert cool tests 🔥

File: src/cow-react/common/utils/isOrderCancellable.test.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+import { NATIVE_CURRENCY_BUY_TOKEN } from 'custom/constants'
+import { isOrderCancellable } from './isOrderCancellable'
+import { COW } from 'constants/tokens'
+
+describe('isOrderCancellable', () => {
+  it('When order cancellation in progress, the it cannot be cancelled', () => {
+    const order = {
+      inputToken: COW[1],
+      status: OrderStatus.PENDING,
+      isCancelling: true, // <-----
```
Comment (anxolin) [inline src/cow-react/common/utils/isOrderCancellable.test.ts:11]: hehe, nice comment. Does the trick :)

File: src/cow-react/common/utils/isOrderCancellable.test.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+      cancellationHash: undefined,
+    } as Order
+
+    expect(isOrderCancellable(order)).toBe(false)
+    expect(isOrderCancellable({ ...order, isCancelling: false })).toBe(true)
```
Comment (anxolin) [inline src/cow-react/common/utils/isOrderCancellable.test.ts:16]: nit: not sure why you test the negative too inside this test. In principle, that should be other test

File: src/cow-react/modules/account/containers/Transaction/StatusDetails.tsx [hunk 5]
```diff
@@ -8,12 +9,14 @@ import OrderCancelledImage from 'assets/cow-swap/order-cancelled.svg'
...
 import OrderOpenImage from 'assets/cow-swap/order-open.svg'
 
-import { StatusLabel, StatusLabelWrapper, StatusLabelBelow } from './styled'
+import { StatusLabel, StatusLabelWrapper, StatusLabelBelow, CancelTxLink } from './styled'
 import { ActivityDerivedState, determinePillColour } from './index'
 import { getSafeWebUrl } from '@cow/api/gnosisSafe'
 import { SafeMultisigTransactionResponse } from '@gnosis.pm/safe-service-client'
 import { getActivityState } from 'hooks/useActivityDerivedState'
 import { CancelButton } from '@cow/common/pure/CancelButton'
+import { ExplorerDataType, getExplorerLink } from '@src/utils/getExplorerLink'
```
Comment (anxolin) [inline src/cow-react/modules/account/containers/Transaction/StatusDetails.tsx:18]: do you need the `@src`?

**Review by anxolin — APPROVED** [review]
This is a great great PR :)

File: src/cow-react/common/hooks/useCancelOrder/useSendOnChainCancellation.test.ts (hunk 1 above)
Comment (shoom3301) [reply src/cow-react/common/hooks/useCancelOrder/useSendOnChainCancellation.test.ts:13]: Ha ha :) Yes, why not?

File: src/cow-react/modules/account/containers/Transaction/StatusDetails.tsx (hunk 5 above)
Comment (shoom3301) [reply src/cow-react/modules/account/containers/Transaction/StatusDetails.tsx:18]: Nope, deleted it, thanks!

**Comment by shoom3301** [comment]
@elena-zh thanks!
1. Could you open an issue for the case please?
2. Fixed
3. See p1
4. I hide it when an order in "open" state
5. Fixed, thanks for the reference!
6. Fixed

File: src/cow-react/common/utils/isOrderCancellable.test.ts (hunk 4 above)
Comment (shoom3301) [reply src/cow-react/common/utils/isOrderCancellable.test.ts:16]: Corresponding to the best practices - yes, it should. But the current version is two times shorter and the readability is not worse, isn't?

File: src/cow-react/modules/account/containers/Transaction/StatusDetails.tsx [hunk 6]
```diff
@@ -8,17 +9,18 @@ import OrderCancelledImage from 'assets/cow-swap/order-cancelled.svg'
...
 import { SafeMultisigTransactionResponse } from '@gnosis.pm/safe-service-client'
 import { getActivityState } from 'hooks/useActivityDerivedState'
 import { CancelButton } from '@cow/common/pure/CancelButton'
+import { ExplorerDataType, getExplorerLink } from 'utils/getExplorerLink'
+import { isOrderCancellable } from '@cow/common/utils/isOrderCancellable'
 
 export function GnosisSafeLink(props: {
   chainId: number
   safeTransaction?: SafeMultisigTransactionResponse
-  gnosisSafeThreshold: number
```
Comment (alfetopito) [inline src/cow-react/modules/account/containers/Transaction/StatusDetails.tsx:21]: Was this not used at all?

**Review by alfetopito — APPROVED** [review]
Very nice!

A few questions on the code

File: src/cow-react/modules/account/containers/Transaction/StatusDetails.tsx (hunk 6 above)
Comment (shoom3301) [reply src/cow-react/modules/account/containers/Transaction/StatusDetails.tsx:21]: Yes

File: src/cow-react/common/hooks/useCancelOrder/useSendOnChainCancellation.ts [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+        return
+      }
+
+      const { sendTransaction } = await getOnChainCancellation(order)
+
+      const receipt = await sendTransaction()
+
+      if (receipt?.hash) {
+        cancelPendingOrder({ id: order.id, chainId })
+        addTransaction({ hash: receipt.hash, ethFlow: { orderId: order.id, subType: 'cancellation' } })
```
[the remark this answers is not in the document]
Comment (shoom3301) [reply src/cow-react/common/hooks/useCancelOrder/useSendOnChainCancellation.ts]: Nice catch! Thank you! Added a condition

**Comment by elena-zh** [comment]
Hey @shoom3301 , 
for the cases 1 and 3 I opened #2235 issue
Cases 2, 4 LGTM!
Cases 5 and 6 are still reproducible to me. 
Could you please take a look at them once again? 

Thanks

**Comment by shoom3301** [comment]
@elena-zh 5,6 cases work well to me.
I guess it's a cache problem. Please, clear your localStorage and try again with new orders.

File: src/custom/state/orders/updaters/GpOrdersUpdater.ts [hunk 8]
```diff
@@ -103,9 +103,10 @@ function _transformGpOrderToStoreOrder(
   // That's why it's not used before and an empty string is set instead
   storeOrder.summary = computeOrderSummary({ orderFromStore: storeOrder, orderFromApi: order }) || ''
 
-  // It can happen that a cancellation is identified in the app before the API is aware
+  // EthFlow adjustments
+  // It can happen that EthFlow cancellation is identified in the app before the API is aware
   // In that case
-  if (order.status === 'cancelled') {
+  if (order.ethflowData && order.status === 'cancelled') {
```
Comment (alfetopito) [inline src/custom/state/orders/updaters/GpOrdersUpdater.ts]: Didn't work?

Comment (shoom3301) [reply src/custom/state/orders/updaters/GpOrdersUpdater.ts]: It changed the behavior for off-chain cancelled orders :(

**Pull request merged**